File: bench/programs/interfaces/interfaces.go
package main

import (
	"fmt"
	"time"
)

type Shape interface {
	area() int64
}

type Circle struct{ r int64 }
type Square struct{ s int64 }

func (c Circle) area() int64 { return c.r * c.r * 355 / 113 }
func (s Square) area() int64 { return s.s * s.s }

func run() int64 {
	var k, rounds, p int64 = 2000, 5000, 1000000007

	shapes := make([]Shape, 0, k)
	for i := int64(0); i < k; i++ {
		v := i%100 + 1
		if i%2 == 0 {
			shapes = append(shapes, Circle{r: v})
		} else {
			shapes = append(shapes, Square{s: v})
		}
	}

	var acc int64 = 0
	for r := int64(0); r < rounds; r++ {
		for _, sh := range shapes {
			acc = (acc + sh.area()) % p
		}
	}
	return acc
}

func main() {
	run() // warmup
	t0 := time.Now()
	ans := run()
	dt := time.Since(t0).Nanoseconds()
	fmt.Println(ans)
	fmt.Println(dt)
}

File: bench/programs/interfaces/interfaces.py
import time


class Circle:
    __slots__ = ("r",)

    def __init__(self, r):
        self.r = r

    def area(self):
        return self.r * self.r * 355 // 113


class Square:
    __slots__ = ("s",)

    def __init__(self, s):
        self.s = s

    def area(self):
        return self.s * self.s


def run():
    k, rounds, p = 2000, 5000, 1000000007
    shapes = []
    for i in range(k):
        v = i % 100 + 1
        shapes.append(Circle(v) if i % 2 == 0 else Square(v))
    acc = 0
    for _ in range(rounds):
        for sh in shapes:
            acc = (acc + sh.area()) % p
    return acc


run()  # warmup
t0 = time.perf_counter_ns()
ans = run()
dt = time.perf_counter_ns() - t0
print(ans)
print(dt)
